File: src/main/java/com/merricklabs/portlandbus/models/ArrivalListPronouncer.java
package com.merricklabs.portlandbus.models;

import com.merricklabs.portlandbus.external.trimet.models.Arrival;
import java.time.LocalDateTime;
import java.util.List;

import static java.util.stream.Collectors.toList;

public class ArrivalListPronouncer {

    private List<ArrivalPronouncer> arrivalPronouncers;
    private int stopId;
    private static final LocalDateTime now = LocalDateTime.now();
    private static final int MAX_ARRIVAL_MINUTES = 30;

    public ArrivalListPronouncer(int stopId, List<Arrival> arrivals) {

        // Filter out arrivals that are in the past and below a time threshold
        this.arrivalPronouncers = arrivals.stream()
                .filter(a -> a.getTime().isAfter(now))
                .filter(a -> a.getMinutesRemaining(now) < MAX_ARRIVAL_MINUTES)
                .map(ArrivalPronouncer::new)
                .collect(toList());
        this.stopId = stopId;
    }

    public static String pronounceStop(int stopId){
        if(stopId <= 1000){
            return String.valueOf(stopId);
        } else {
            return "<say-as interpret-as=\"digits\">" + stopId + "</say-as>";
        }
    }

    private String pronounceStop(){
        return pronounceStop(stopId);
    }

    public String pronounceArrivals(){
        final LocalDateTime now = LocalDateTime.now();

        if (arrivalPronouncers.size() == 0){
            return "No arrivals are currently scheduled for stop " + pronounceStop() + " in the next " + MAX_ARRIVAL_MINUTES + " minutes.";
        }

        StringBuilder speechBuilder = new StringBuilder("Next arrivals at stop " + pronounceStop() + ": ");

        if (arrivalPronouncers.size() == 1){
            speechBuilder.append(arrivalPronouncers.get(0).pronounceForMultipleArrival(now));
        } else {
            arrivalPronouncers.subList(0, arrivalPronouncers.size() - 1)
                    .stream()
                    .forEach(a -> speechBuilder.append(a.pronounceForMultipleArrival(now) + ", "));

            speechBuilder.append("and " + arrivalPronouncers.get(arrivalPronouncers.size() - 1).pronounceForMultipleArrival(now) + ".");
        }
        return speechBuilder.toString();
    }

    /**
     * Format arrivals for display on Echo Show
     * @return
     */
    public String showArrivals(){
        StringBuilder textBuilder = new StringBuilder("Stop " + stopId + ": \n");

        if (arrivalPronouncers.size() == 0){
            textBuilder.append("No arrivals");
        } else {
            arrivalPronouncers
                    .stream()
                    .forEach(a -> textBuilder.append(a.showNextArrival(now) + "\n"));
        }
        return textBuilder.toString();
    }
}

File: src/main/kotlin/com/merricklabs/portlandbus/models/ArrivalListPronouncer.kt
package com.merricklabs.portlandbus.models

import com.merricklabs.portlandbus.external.trimet.models.Arrival
import java.time.LocalDateTime

class ArrivalListPronouncer(private val stopId: Int, arrivals: List<Arrival>) {

    private val arrivalPronouncers: List<ArrivalPronouncer>

    init {
        // Filter out arrivals that are in the past and below a time threshold
        this.arrivalPronouncers = arrivals
                .asSequence()
                .filter { it.time.isAfter(now) }
                .filter { it.getMinutesRemaining(now) < MAX_ARRIVAL_MINUTES }
                .map { ArrivalPronouncer(it) }
                .toList()
    }

    private fun pronounceStop(): String {
        return pronounceStop(stopId)
    }

    fun pronounceArrivals(): String {
        val now = LocalDateTime.now()

        if (arrivalPronouncers.isEmpty()) {
            return "No arrivals are currently scheduled for stop ${pronounceStop()} in the next $MAX_ARRIVAL_MINUTES minutes."
        }

        val speechBuilder = StringBuilder("Next arrivals at stop ${pronounceStop()}: ")

        if (arrivalPronouncers.size == 1) {
            speechBuilder.append(arrivalPronouncers[0].pronounceForMultipleArrival(now))
        } else {
            arrivalPronouncers.subList(0, arrivalPronouncers.size - 1)
                    .forEach { speechBuilder.append("${it.pronounceForMultipleArrival(now)}, ") }

            speechBuilder.append("and ${arrivalPronouncers[arrivalPronouncers.size - 1].pronounceForMultipleArrival(now)}.")
        }
        return speechBuilder.toString()
    }

    /**
     * Format arrivals for display on Echo Show
     * @return
     */
    fun showArrivals(): String {
        val textBuilder = StringBuilder("Stop $stopId: \n")

        if (arrivalPronouncers.isEmpty()) {
            textBuilder.append("No arrivals")
        } else {
            arrivalPronouncers
                    .stream()
                    .forEach { textBuilder.append("${it.showNextArrival(now)}\n") }
        }
        return textBuilder.toString()
    }

    companion object {
        private val now = LocalDateTime.now()
        private val MAX_ARRIVAL_MINUTES = 30

        fun pronounceStop(stopId: Int): String {
            return if (stopId <= 1000) {
                stopId.toString()
            } else {
                "<say-as interpret-as=\"digits\">$stopId</say-as>"
            }
        }
    }
}
